### ISSUE TITLE
Cannot retrieve error info after an invalid request is made

### ISSUE THREAD

**Issue by russgove** [SharePoint/PnP-JS-Core#274]
Thank you for reporting an issue, suggesting an enhancement, or asking a question. We appreciate your feedback - to help the team understand your
needs please complete the below template to ensure we have the details to help. Thanks!

#### Category
[x] Enhancement

[ ] Bug

[ x] Question

I have code that lets the user enter the fields he wats to select from a list, and then retrives the data and shows it.  the code is as follows
        const promise = web.lists.getById(listid).items.select(fieldnames.concat("GUID").concat("Id").join(",")).get()
            .then((response) => {

                const data = _.map(response, (item: any) => {
                    item.__metadata__ListDefinitionId = listDefinition.guid; // save my listdef, so i can get the columnReferences later
                    return item;
                });
                console.log(data);
                const gotListItems = gotListItemsAction(data);
                dispatch(gotListItems); // need to ewname this one to be digfferent from the omported ome
            })
            .catch((error) => {
                console.log(error);
                dispatch(getListItemsErrorAction(error)); // need to ewname this one to be digfferent from the omported ome
            });
        promises.push(promise);

If the user selects an invalid field name I end up in the catch block. The errror  object in the catch block looks like this
action.payload.error
TypeError: Already read
    at ODataDefaultParser.ODataParserBase.parse (odata.js:33)
    at Items.Queryable.processHttpClientResponse (queryable.js:309)
    at queryable.js:242

The request that got sent to the server was
https://xxxx.sharepoint.com/sites/issues/Ussuessite1/_api/web/lists('e06e5b0c-4734-487a-867c-9973e5e915f9')/items?$select=Title,_UIVersionString,GUID,Id

and the response received was 
HTTP/1.1 400 Bad Request
Cache-Control: private, max-age=0
Transfer-Encoding: chunked
Content-Type: application/json;odata=minimalmetadata;streaming=true;charset=utf-8
Expires: Mon, 05 Dec 2016 15:19:57 GMT
Last-Modified: Tue, 20 Dec 2016 15:19:57 GMT
Server: Microsoft-IIS/8.5

X-SharePointHealthScore: 0
X-SP-SERVERSTATE: ReadOnly=0
DATASERVICEVERSION: 3.0
SPClientServiceRequestDuration: 38
SPRequestDuration: 294
X-AspNet-Version: 4.0.30319
SPRequestGuid: e2efc29d-f0af-3000-d0ff-95171178acb7
request-id: e2efc29d-f0af-3000-d0ff-95171178acb7
Strict-Transport-Security: max-age=31536000
X-FRAME-OPTIONS: SAMEORIGIN
X-Powered-By: ASP.NET
MicrosoftSharePointTeamServices: 16.0.0.6008
X-Content-Type-Options: nosniff
X-MS-InvokeApp: 1; RequireReadOnly
P3P: CP="ALL IND DSP COR ADM CONo CUR CUSo IVAo IVDo PSA PSD TAI TELo OUR SAMo CNT COM INT NAV ONL PHY PRE PUR UNI"
Date: Tue, 20 Dec 2016 15:19:57 GMT

9d
{"odata.error":{"code":"-1, Microsoft.SharePoint.SPException","message":{"lang":"en-US","value":"The field or property '_UIVersionString' does not exist."}}}
0

I  need to be able to capture the error tect "The field or property '_UIVersionString' does not exist." so othat I can display it to the user.

I traced it through and it looks like Queryable.prototype.processHttpClientResponse  calls parser.parse(response); and that does not return the values properly.

I think it might be because a few lines up in processHttpClientResponse we are calling response.text() to get the error message and log it, and the message cannot be read twice from the fetch client.

**Comment by tavikukko** [comment]
Found the same issue, reported here -> #270 . @patrick-rodgers got fix for that, not yet checked in thou.

**Comment by patrick-rodgers** [comment]
Yes, this bug is fixed and will be included in the 2.0 release. Issue was not returning after throwing the error within the promise (which allowed the current code to execute). So now everything is handled in a single returned promise for simplicity. Going to leave this open for now in case others report it so we can refer until we get it merged.

**Comment by patrick-rodgers** [comment]
If you get a chance, please test you scenario with the dev branch so we can ensure it is fixed as expected. Thanks! Dev branch is currently what will be 2.0

**Comment by russgove** [comment]
I downloaded the dev branch to a new folder, ran npm install, and then gulp serve. I got this error:
D:\github3\PNP-JS-Core-V2 (master) (sp-pnp-js@2.0.0)                                                                  
λ gulp serve                                                                                                          
D:\github3\PNP-JS-Core-V2\gulptasks\@configuration.js:5                                                               
    let pkg = require("../package.json");                                                                             
    ^^^                                                                                                               
                                                                                                                      
SyntaxError: Block-scoped declarations (let, const, function, class) not yet supported outside strict mode            
    at exports.runInThisContext (vm.js:53:16)                                                                         
    at Module._compile (module.js:373:25)                                                                             
    at Object.Module._extensions..js (module.js:416:10)                                                               
    at Module.load (module.js:343:32)                                                                                 
    at Function.Module._load (module.js:300:12)                                                                       
    at Module.require (module.js:353:17)                                                                              
    at require (internal/module.js:12:17)                                                                             
    at Object.<anonymous> (D:\github3\PNP-JS-Core-V2\gulptasks\clean.js:10:14)                                        
    at Module._compile (module.js:409:26)                                                                             
    at Object.Module._extensions..js (module.js:416:10)                                                               
                                                                                                                      
D:\github3\PNP-JS-Core-V2 (master) (sp-pnp-js@2.0.0)                                                                  
λ dir        



Did i miss a step????

**Comment by tavikukko** [comment]
Dev Branch does not build. Tried to fix by adding "use strict" to @configuration.js, build.js, test.js, serve.js, sync.js but there were also errors in digestcache.ts, nodefetchclient.ts, items.ts and so on. There is something wrong with this branch, some imports might be missing, didn't have time to look closer. @patrick-rodgers should take fresh clone of this branch and see whats going on. Happy holidays :)

**Comment by patrick-rodgers** [comment]
You'll need to upgrade to node 6.9.* which is the latest LTS version. Includes support for many es6 features. What version of node are you two running? Have mentioned this on several of the calls but haven't written up the migration article yet for wiki.

**Comment by tavikukko** [comment]
Upgraded node to 6.9.1 but still some type errors

**src/net/digestcache.ts(40,42):** error TS2339: Property 'FormDigestValue' does not exist on type 'Response'.
**src/net/digestcache.ts(41,32):** error TS2339: Property 'FormDigestTimeoutSeconds' does not exist on type 'Response'.
**src/net/nodefetchclient.ts(55,16):** error TS2322: Type 'Promise<string>' is not assignable to type 'Promise<AuthToken>'.
  Type 'string' is not assignable to type 'AuthToken'.
**src/sharepoint/items.ts(177,16):** error TS2322: Type 'Promise<{ ListItemEntityTypeFullName: string; }>' is not assignable to type 'Promise<ItemUpdateResult>'.
  Type '{ ListItemEntityTypeFullName: string; }' is not assignable to type 'ItemUpdateResult'.
    Property 'item' is missing in type '{ ListItemEntityTypeFullName: string; }'.
**src/sharepoint/userprofiles.ts(166,16):** error TS2322: Type 'Promise<ArrayBuffer>' is not assignable to type 'Promise<void>'.
  Type 'ArrayBuffer' is not assignable to type 'void'.

**Comment by patrick-rodgers** [comment]
It appears that there was a change to the TS transpiler between when I did the initial check-in and now. Those issues did not appear until I update TS. What is strange is mostly it seems to no longer be able to resolve the child promise returns correctly and is creating the error based on the first promise in the chain's return type. I am making updates to deal with this and will push them in soon.

**Comment by patrick-rodgers** [comment]
updates pushed

**Comment by russgove** [comment]
if I upgrade to node 6.9.1 should I expect my old code repos to work?

**Comment by russgove** [comment]
OK, I upgraded t0 node 6.9.2 and got the latest from the dev branch.
gulp package ran fine.
Gulp serve threw a bunch of errors:
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\utils\util.test.ts                                                                                                                                           
(8,34): error TS7006: Parameter 'a' implicitly has an 'any' type.                                                                                                                                     
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\utils\util.test.ts                                                                                                                                           
(8,39): error TS2683: 'this' implicitly has type 'any' because it does not have a type annotation.                                                                                                    
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\utils\util.test.ts                                                                                                                                           
(8,50): error TS2683: 'this' implicitly has type 'any' because it does not have a type annotation.                                                                                                    
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\utils\storage.test.ts                                                                                                                                        
(31,72): error TS6133: 'r' is declared but never used.                                                                                                                                                
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\utils\storage.test.ts                                                                                                                                        
(31,75): error TS6133: 'n' is declared but never used.                                                                                                                                                
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\utils\storage.test.ts                                                                                                                                        
(2,26): error TS1149: File name 'D:\github3\PNP-JS-Core-V2\src\utils\Storage.ts' differs from already included file name 'D:/github3/PNP-JS-Core-V2/src/utils/storage.ts' only in casing              
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\utils\logging.test.ts                                                                                                                                        
(37,60): error TS6133: 'e' is declared but never used.                                                                                                                                                
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\utils\logging.test.ts                                                                                                                                        
(38,60): error TS6133: 'e' is declared but never used.                                                                                                                                                
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\utils\logging.test.ts                                                                                                                                        
(39,60): error TS6133: 'e' is declared but never used.                                                                                                                                                
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\utils\logging.test.ts                                                                                                                                        
(45,47): error TS6133: 'e' is declared but never used.                                                                                                                                                
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\utils\logging.test.ts                                                                                                                                        
(46,47): error TS6133: 'e' is declared but never used.                                                                                                                                                
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\utils\logging.test.ts                                                                                                                                        
(47,47): error TS6133: 'e' is declared but never used.                                                                                                                                                
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\test-config.test.ts                                                                                                                                          
(3,13): error TS6133: 'mocha' is declared but never used.                                                                                                                                             
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\test-config.test.ts                                                                                                                                          
(71,19): error TS7006: Parameter 'element' implicitly has an 'any' type.                                                                                                                              
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\test-config.test.ts                                                                                                                                          
(74,51): error TS6133: 'index' is declared but never used.                                                                                                                                            
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\test-config.test.ts                                                                                                                                          
(74,58): error TS6133: 'arr' is declared but never used.                                                                                                                                              
                                                                                                                                                                                                      
ERROR in D:\github3\PNP-JS-Core-V2\tests\sharepoint\topnavigationbar.test.ts                                                                                                                          
(2,34): error TS1149: File name 'D:\github3\PNP-JS-Core-V2\src\sharepoint\topNavigationBar.ts' differs from already included file name 'D:/github3/PNP-JS-Core-V2/src/sharepoint/topnavigationbar.ts' 
only in casing

**Comment by russgove** [comment]
And after upgrading to node 6.9.2, I went BACK to my old project (an spfx project that users pnp-js-core). I  ran a gulp serve and got an error saying  things had change since I last ran npm install, So I deleted my node_modules directory and ran npm install.
now when I run gulp serve on my old project , all I get is 
D:\github3\spfx-reactgrid2 (dev) (spfxreactgrid@0.0.1)
λ gulp serve
[15:27:29] Error - Unknown
 Path must be a string. Received null
[15:27:29] ==================[ Finished ]==================
[15:27:30] Project spfxreactgrid version: 0.0.1
[15:27:30] Build tools version: 0.11.1
[15:27:30] Node version: v6.9.2
[15:27:30] Total duration: 3.13 s
[15:27:30] Task errors: 2


any idea what I did wrong/how to fix this?

**Comment by tavikukko** [comment]
Seems to build now ok. I tested the new exception handler, something needs to be fixed:

run this:
```
$pnp.sp.web.lists.getByTitle("ImNotHere").get().then(function(result) {
    console.log(result);
}).catch(function(error) {
    console.log(error); // --> TypeError: _this2.Log is not a function
});`
```

I would still prefer that the exception handler would return json object, using response.json() method, which returns much more information about the error.

So now (when we get rid of this **_this2.Log is not a function** error), when we receive an error,
we still need to read that json from that error response:

```
pnp.sp.web.lists.getByTitle("ImNotHere").get().then(function (result) {
   console.log(result);
}).catch(function (error) {
   error.response.json().then((json) => {
      console.log(json["odata.error"].message.value);
      // will print:
      // List 'ImNotHere' does not exist at site with URL 'https://contoso.sharepoint.com'.
   });
});
```

In my opinion, we should return that json object, just to make it easier to show underlying error messages which are returned from SharePoint rest api.

**Comment by patrick-rodgers** [comment]
I'll admit this thread is getting hard to follow.

@russgove - I am not sure, you can use [NVM](https://github.com/creationix/nvm) to manage multiple versions of node on your machine. I have tested node 6.9.* with the latest SPFx release but haven't gone back and tried it with all of them so I am not sure what you might have run into. I do now see the errors when you run gulp serve and am working to clean them up, thanks for pointing them out.

@tavikukko - I think there is value in returning a custom exception so plan to stick with that. You can for example make decisions in the catch based on the type. I have updated the exception to extract the json() value and use that in the data value instead of the entire raw request as I agree that makes more sense. The model we are encouraging is using the logger listeners instead of hitting console directly which does get you a more succinct message than the entire exception object, which I think is your complaint. I was not able to duplicate your issue with the not a function error, but did update the calls to Log to super from this. The below is what is logged to the logger:

```
Message: Error making HttpClient request in queryable: [404] Not Found. Data: {"odata.error":{"code":"-1, System.ArgumentException","message":{"lang":"en-US","value":"List 'asdfasdfa' does not exist at site with URL 'https://318studios.sharepoint.com/sites/dev'."}}}
```

**Comment by tavikukko** [comment]
Ok, I was not aware of the logger listeners, will start using that now. Tried to find good sample how to use the logger but no luck. How do I get that message logged in my console using logger listeners?

**Comment by tavikukko** [comment]
ok, got it now.

```
import * as pnp from "pnp";
import { Logger, ConsoleListener, LogLevel } from 'utils/logging';

Logger.clearSubscribers();
Logger.subscribe(new ConsoleListener());

pnp.sp.web.lists.getByTitle('NotHere').get().then(result => {
   console.log(result);
});
```

**Comment by patrick-rodgers** [comment]
Can we close this issue as the original bug is fixed. If there are other bugs new issues can be opened, but seems like what this was originally opened for is now fixed. Thoughts?

**Comment by tavikukko** [comment]
Yes, the bug is fixed. There was also need to return the fetch response error to the client, it is very useful in the **catch block** and i think it was the original request from @russgove to get that error and to be able display it to the user. But lets close this and open new issue, if this needs to be discussed more.

**Comment by patrick-rodgers** [comment]
Ok, going to close this. I have updated the error to include the json body in the object so that console.log(e) will work as folks expect. That should meet the original ask as well as address the other points brought up in the thread. Change is in dev branch.

**Comment by tavikukko** [comment]
tested and works well, thanks!

**Comment by tavikukko** [comment]
Just realised that this does not work.. has something changed between 2.0.0 --> 2.0.2?

`
pnp.sp.web.lists.getByTitle('NotHere').get().then(result => {
   console.log(result);
})
.catch(e => {
	console.log(e);
});
`

outputs this:
"Error: Error making HttpClient request in queryable: [404] Not Found"
